core/src/apps/cardano/sign_tx.py

```diff
@@ -96,44 +99,52 @@ async def sign_tx(
 
     validate_network_info(msg.network_id, msg.protocol_magic)
 
-    if _has_stake_pool_registration(msg):
-        return await _sign_stake_pool_registration_tx(ctx, msg, keychain)
-    else:
-        return await _sign_standard_tx(ctx, msg, keychain)
-
-
-async def _sign_standard_tx(
-    ctx: wire.Context, msg: CardanoSignTx, keychain: seed.Keychain
-) -> CardanoSignedTx:
     try:
-        for i in msg.inputs:
-            await validate_path(
-                ctx, keychain, i.address_n, SCHEMA_ADDRESS.match(i.address_n)
+        if _has_stake_pool_registration(msg):
+            cborized_tx, tx_hash = await _sign_stake_pool_registration_tx(
+                ctx, msg, keychain
             )
+        else:
+            cborized_tx, tx_hash = await _sign_ordinary_tx(ctx, msg, keychain)
 
-        _validate_outputs(keychain, msg.outputs, msg.protocol_magic, msg.network_id)
-        _validate_certificates(msg.certificates, msg.protocol_magic, msg.network_id)
-        _validate_withdrawals(msg.withdrawals)
-        _validate_metadata(msg.metadata)
+        signed_tx_chunks = _get_signed_tx_chunks(cborized_tx, tx_hash)
 
-        # display the transaction in UI
-        await _show_standard_tx(ctx, keychain, msg)
+        for tx_chunk_message in signed_tx_chunks:
+            if tx_chunk_message.expect_more_chunks:
+                await ctx.call(tx_chunk_message, CardanoSignedTxAck)
+            else:
+                return tx_chunk_message
 
-        # sign the transaction bundle and prepare the result
-        serialized_tx, tx_hash = _serialize_tx(keychain, msg)
-        tx = CardanoSignedTx(serialized_tx=serialized_tx, tx_hash=tx_hash)
+        raise AssertionError("Should not be reached")
 
     except ValueError as e:
         if __debug__:
             log.exception(__name__, e)
         raise wire.ProcessError("Signing failed")
 
-    return tx
+
+async def _sign_ordinary_tx(
+    ctx: wire.Context, msg: CardanoSignTx, keychain: seed.Keychain
+) -> Tuple[CborizedSignedTx, TxHash]:
+    for i in msg.inputs:
+        await validate_path(
```

Comment:
I know this is the original code but why is validate_path async?

Reply:
it's a common (across coins) helper function that as part of the validation potentially triggers a prompt warning the user that given path is not standard: https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/common/paths.py#L250